File: app/src/main/java/app/trigger/BluetoothDoorSetup.java
package app.trigger;

import android.graphics.Bitmap;

import app.trigger.DoorState.StateCode;


public class BluetoothDoorSetup implements Setup {
    static final String type = "BluetoothDoorSetup";
    int id;
    String name;
    public String device_name;
    public String service_uuid;
    public String open_query;
    public String close_query;
    public String ring_query;
    public String locked_pattern;
    public String unlocked_pattern;
    public String status_query;
    public Bitmap open_image;
    public Bitmap closed_image;
    public Bitmap unknown_image;
    public Bitmap disabled_image;

    public BluetoothDoorSetup(int id, String name) {
        this.id = id;
        this.name = name;
        this.device_name = "";
        this.service_uuid = "";
        this.open_query = "";
        this.close_query = "";
        this.ring_query = "";
        this.locked_pattern = "LOCKED";
        this.unlocked_pattern = "UNLOCKED";
        this.status_query = "";
        this.open_image = null;
        this.closed_image = null;
        this.unknown_image = null;
        this.disabled_image = null;
    }

    @Override
    public int getId() {
        return id;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getType() {
        return type;
    }

    @Override
    public String getWiFiSSIDs() {
        return device_name;
    }

    @Override
    public boolean getWiFiRequired() {
        return false;
    }

    @Override
    public Bitmap getStateImage(StateCode state) {
        switch (state) {
            case OPEN:
                return open_image;
            case CLOSED:
                return closed_image;
            case DISABLED:
                return disabled_image;
            case UNKNOWN:
                return unknown_image;
            default:
                return null;
        }
    }

    @Override
    public String getRegisterUrl() {
        return "";
    }

    @Override
    public DoorState parseReply(DoorReply reply) {
        return Utils.genericDoorReplyParser(reply, this.unlocked_pattern, this.locked_pattern);
    }

    @Override
    public boolean canOpen() {
        return !Utils.isEmpty(open_query);
    }

    @Override
    public boolean canClose() {
        return !Utils.isEmpty(close_query);
    }

    @Override
    public boolean canRing() {
        return !Utils.isEmpty(ring_query);
    }
}

File: app/src/main/kotlin/app/trigger/BluetoothDoorSetup.kt
package app.trigger

import android.graphics.Bitmap
import app.trigger.DoorState.StateCode
import org.json.JSONObject


class BluetoothDoorSetup(override var id: Int, override var name: String) : Setup {
    override val type = Companion.TYPE
    var device_name = ""
    var service_uuid = ""
    var open_query = ""
    var close_query = ""
    var ring_query = ""
    var locked_pattern = "LOCKED"
    var unlocked_pattern = "UNLOCKED"
    var status_query = ""
    var open_image: Bitmap? = null
    var closed_image: Bitmap? = null
    var unknown_image: Bitmap? = null
    var disabled_image: Bitmap? = null

    override fun getWiFiSSIDs(): String =  ""
    override fun getWiFiRequired(): Boolean = false

    override fun getStateImage(state: StateCode?): Bitmap? {
        return when (state) {
            StateCode.OPEN -> open_image
            StateCode.CLOSED -> closed_image
            StateCode.DISABLED -> disabled_image
            StateCode.UNKNOWN -> unknown_image
            else -> null
        }
    }

    override fun parseReply(reply: DoorReply): DoorState {
        return Utils.genericDoorReplyParser(reply, unlocked_pattern, locked_pattern)
    }

    override fun canOpen(): Boolean {
        return !Utils.isEmpty(open_query)
    }

    override fun canClose(): Boolean {
        return !Utils.isEmpty(close_query)
    }

    override fun canRing(): Boolean {
        return !Utils.isEmpty(ring_query)
    }

   fun toJSONObject(): JSONObject {
        val obj = JSONObject()
        obj.put("id", id)
        obj.put("name", name)
        obj.put("type", type)

        obj.put("device_name", device_name)
        obj.put("service_uuid", service_uuid)
        obj.put("open_query", open_query)
        obj.put("close_query", close_query)
        obj.put("ring_query", ring_query)
        obj.put("locked_pattern", locked_pattern)
        obj.put("unlocked_pattern", unlocked_pattern)
        obj.put("status_query", status_query)

        obj.put("open_image", Utils.serializeBitmap(open_image))
        obj.put("closed_image", Utils.serializeBitmap(closed_image))
        obj.put("unknown_image", Utils.serializeBitmap(unknown_image))
        obj.put("disabled_image", Utils.serializeBitmap(disabled_image))

        return obj
    }

    companion object {
        const val TYPE = "BluetoothDoorSetup"

        fun fromJSONObject(obj: JSONObject): BluetoothDoorSetup {
            val id = obj.getInt("id")
            val name = obj.getString("name")
            val setup = BluetoothDoorSetup(id, name)

            setup.device_name = obj.optString("device_name", "")
            setup.service_uuid = obj.optString("service_uuid", "")
            setup.open_query = obj.optString("open_query", "")
            setup.close_query = obj.optString("close_query", "")
            setup.ring_query = obj.optString("ring_query", "")
            setup.locked_pattern = obj.optString("locked_pattern", "LOCKED")
            setup.unlocked_pattern = obj.optString("unlocked_pattern", "UNLOCKED")
            setup.status_query = obj.optString("status_query", "")

            setup.open_image = Utils.deserializeBitmap(obj.optString("open_image", ""))
            setup.closed_image = Utils.deserializeBitmap(obj.optString("closed_image", ""))
            setup.unknown_image = Utils.deserializeBitmap(obj.optString("unknown_image", ""))
            setup.disabled_image = Utils.deserializeBitmap(obj.optString("disabled_image", ""))

            return setup
        }
    }
}
